### DISCUSSION TITLE
Request for a Roll-back to previous C# version analyzer

### DISCUSSION THREAD

**Discussion by markpahulje** [dotnet/roslyn#49437] [Ideas]
In general, some newer C# features include short forms expression of a longer piece of a code fragment or boiler-template. I am requesting a Roll-Back Code Version Analyzer that would allow that would expand the short form to it full code. It should automatically pick, the previous rollback C# language version ( C#?.0 ), since that's when the short form was started introduced. For example, a short form C#9.0 feature would be to rolled-back to C# 6.0 long version of the code. It be great to see it go back to c#1.0 but I would settle for C#4.0 as a start of the base C# version to begin with. But it would be really cool to see the that chain of decisions in reverse, reflected in code. I really think this would help developers to because it would readily surface the history of the design decisions of the language, which now you have to glean from forms and videos. Hope you all agree. Let's make this happen.

Take for example auto-property; where I would like to get expansion of 

public string Name { get; set; } = "bunny"

hints and can replace back to 

private string name = "bunny"; 
public string Name
{
    get
    {
        return http://this.name;
    }
    set
    {
        http://this.name = value;
    }
}

This would be a separate analyzer.

**Comment by PathogenDavid** [comment]
Not an analyzer, but if your main focus is education and intrigue you can basically get this with [SharpLab's C# decompilation mode](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQZgAJ1CBhXAb10JuKJQwAZCA5CMOQiwgczhgDchKPyEBfQgF5CAImABXAHaKAnjIG4xQA==). If you want more granular application of C# features you can use [ILSpy](https://github.com/icsharpcode/ILSpy) (which is what SharpLab is using on the backend, but it exposes more configuration.)

**Reply by markpahulje** [reply]
ILSpy is very good as decompiler, agreed. But sometimes you want to understand what to get back to basic and get full expansion of the code short-cut. Let's take a very simple auto-property, well if I want to do something interesting with it then I want the full expansion of what that. 

For example, I want to multiple my setter with Pie.  I want to expand 
public int X { get; set; } = 1; 

//Short-form  public int X { get; set; }=initialvalue - full expansion introduced in C#6.0
private int x = 1;
public  int X
{
get
{
return this.x;
}
set
{
x = value*3.1459; //then I can multiple by pie 
}
}

The idea is I want to set the expansion to a previous version of C# version where the short-cut first took place and see the chain of changes as relate to that short-cut code. 

public int X { get; set; };
then  
public int X { get; set; }=1; 

revert back to their respective full expansions. Each has a c# version when they were introduced.

**Reply by PathogenDavid** [reply]
> But sometimes you want to understand what to get back to basic and get full expansion of the code short-cut.

Yeah, I understand what you're asking for. ILSpy lets you do this, just disable the language versions you don't want it to use:

![image](https://user-images.githubusercontent.com/278957/99731140-1537e880-2a83-11eb-9256-7003f4373cd5.png)

**Reply by markpahulje** [reply]
Oh I did not know that. Nice. But I don't want to get out of VS Code/VS Studio to do that. Now they have the source code to port easily enough!

**Comment by ufcpp** [comment]
Convert to full property?

![image](https://user-images.githubusercontent.com/3760955/99474337-ba719480-298f-11eb-998e-7e73b783950b.png)

**Reply by markpahulje** [reply]
Yes, it's a start but you want to select the target C# level to revert to. So say I want to revert to C#4.0 so the auto-property fully explode to this; 

//Short-form  public int X { get; set; } full expansion as introduced in C#5.0 
private int x = defaultvalue;
public int  X 
{
  get
  {
   return this.x; 
  }
  set
  {
    x = value;
  }
}
Of course this is just one example of a short form, not sure when it was introduced c#5.0?, but it continues now.

**Comment by CyrusNajmabadi** [comment]
I don't think this fits into Roslyn itself.  It seems like it might be a good community project.

**Comment by svick** [comment]
I don't understand what is the goal here; *why* do you want to rollback the code to how it would look in a previous version of the language?

If it's about education, then, as mentioned, ILSpy is a decent (though not perfect) approach.

If it's about being able to modify your code in a way that can only be done in the older version of the syntax ("I want to multiply my setter with pi"), then language versions are irrelevant, you're just looking for a refactoring. As mentioned, such refactorings already exist. And if they're not sufficient for your needs, or if one is missing, then I think you should open specific issue about a specific refactoring.
